lib/flight_scheduler/task_registry.rb

```diff
@@ -0,0 +1,123 @@
+#==============================================================================
+# Copyright (C) 2020-present Alces Flight Ltd.
+#
+# This file is part of FlightSchedulerController.
+#
+# This program and the accompanying materials are made available under
+# the terms of the Eclipse Public License 2.0 which is available at
+# <https://www.eclipse.org/legal/epl-2.0>, or alternative license
+# terms made available by Alces Flight Ltd - please direct inquiries
+# about licensing to licensing@alces-flight.com.
+#
+# FlightSchedulerController is distributed in the hope that it will be useful, but
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with FlightSchedulerController. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on FlightSchedulerController, please visit:
+# https://github.com/openflighthpc/flight-scheduler-controller
+#==============================================================================
+
+class FlightScheduler::TaskRegistry
+  attr_reader :job
+
+  def initialize(job)
+    @job = job
+    @next_task = task_enum.next
+    @running_tasks = []
+    @past_tasks = []
+    @mutex = Mutex.new
+  end
+
+  def all_tasks(update = true)
+    refresh if update
+    [@next_task, *@running_tasks, *@past_tasks]
+  end
+
+  def next_task(update = true)
+    refresh if update
+    @next_task
+  end
+
+  def running_tasks(update = true)
+    refresh if update
+    @running_tasks
+  end
+
+  def past_tasks(update = true)
+    refresh if update
+    @past_tasks
+  end
+
+  def max_tasks_running?(update = true)
+    refresh if update
+    @running_tasks.length >= job.max_nodes
+  end
+
+  def finished?(update = true)
+    refresh if update
+    if @next_task
+      false
+    else
+      @running_tasks.empty?
+    end
+  end
+
+  private
+
+  def refresh
+    @mutex.synchronize do
+      @running_tasks.select! do |task|
+        # Transition "finalised" tasks from running to past
+        if task.running? || (task.allocated? && task.pending?)
+          true
+        else
+          @past_tasks << task
+          false
+        end
+      end
```

Comment:
This is easier to understand than it was before.  However, I still have some issues with it.

This method is modifying mutable instance variables within the `synchronize` block.  If say, `running_tasks(false)` or `past_tasks(false)` are called whilst this method is running in some other thread, they will get the partially updated `@running_tasks`/`@past_tasks`.  This could result in a task appearing twice or not appearing at all.

We can reduce that race condition like this (totally untested code follows):

```ruby
@mutex.synchronize do
  now_running, now_past = @running_tasks.partition { |task| task.running? || (task.allocated? && task.pending?) }
  # Handle next task as before, but using now_running and now_past instead of @running_tasks and @past_tasks
  @running_tasks = now_running
  @past_tasks.concat(now_past)
end
```

Replacing `select!` with `partition` is also easier to understand its purpose.

Reply:
Done: https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/1266e37367c51b8d31741a803c0293acb2a8ebf0

Reply:
And a bug fix + spec: https://github.com/openflighthpc/flight-scheduler-controller/pull/12/commits/1d3b990168e1243edec95b3c3b6e5b14128e9b55

Reply:
This is fine for now.